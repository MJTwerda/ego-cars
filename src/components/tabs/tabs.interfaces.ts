export enum TabValue {
  MODELS = 'models',
  MODELS_TAB = 'modelsTab',
};

export enum TabColors {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
}

export interface TabI {
  value: TabValue;
  label: string | null;
};