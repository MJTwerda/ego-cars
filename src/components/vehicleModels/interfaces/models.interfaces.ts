export interface VehicleModelI {
  id: number;
  name: string;
  segment: string;
  year: number;
  price: number;
  thumbnail: string;
  photo: string;
};

export interface ModelFeaturesI {
  description: string;
  image: string;
  name: string;
};

export interface ModelHighlightsI {
  content: string;
  image: string;
  title: string;
};

export interface VehicleModeldetailsI extends VehicleModelI {
  description: string;
  model_features: ModelFeaturesI[];
  model_highlights: ModelHighlightsI[];
  title: string;
};