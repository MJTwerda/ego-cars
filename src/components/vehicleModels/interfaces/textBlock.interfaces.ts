export interface ImageBlock {
  src: string;
  alt: string;
  position: 'right' | 'left';
};

export interface TextInfoBlock {
  model?: string;
  title: {
    text: string;
    variant: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'subtitle1' | 'subtitle2' | 'body1' | 'body2' | 'caption' | 'button' | 'overline';
  };
  description: string;
};