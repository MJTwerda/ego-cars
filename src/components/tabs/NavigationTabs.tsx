'use client'
import { Tab, Tabs } from "@mui/material";
import { usePathname, useRouter } from "next/navigation";
import { useEffect, useState } from "react";
import { TabColors, NavigationTabI } from './tabs.interfaces';

interface CommonTabsProps {
  tabList: NavigationTabI[];
  textColor?: TabColors;
  indicatorColor?: TabColors;
}

const NavigationTabs = ({ tabList, textColor = TabColors.PRIMARY, indicatorColor = TabColors.PRIMARY }: CommonTabsProps) => {
  const router = useRouter();
  const pathname = usePathname();
  const [ currentTabPath, setCurrentTabPath ] = useState<string>("");

  useEffect(() => {
    const matchingTab = tabList.find(tab => `/${tab.path}` === pathname);
    if (matchingTab) {
      setCurrentTabPath(matchingTab.path);
    } else {
      setCurrentTabPath("/");
    }
  }, [pathname, tabList]);

  const handleNavigateByTab = (e: React.SyntheticEvent, tabPath: string) => {
    setCurrentTabPath(tabPath);
    router.push(`/${tabPath}`);
  };

  return (
    <Tabs 
      value={currentTabPath || false} 
      onChange={handleNavigateByTab} 
      textColor={textColor} 
      indicatorColor={indicatorColor}
    >
      {tabList.map((tab: NavigationTabI) => (
        <Tab 
          key={tab.path} 
          component="a"
          id={`navigation-tab-${tab.path}`}
          value={tab.path} 
          label={tab.label} 
        />
      ))}
    </Tabs>
  );
};

export default NavigationTabs;