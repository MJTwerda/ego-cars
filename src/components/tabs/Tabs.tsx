'use client'
import { Tab, Tabs } from "@mui/material";
import { useState } from "react";
import { TabColors, TabI, TabValue } from './tabs.interfaces';

interface CommonTabsProps {
  tabList: TabI[];
  textColor?: TabColors;
  indicatorColor?: TabColors;
}

const CommonTabs = ({tabList, textColor = TabColors.PRIMARY, indicatorColor = TabColors.PRIMARY}: CommonTabsProps) => {
  const [ tabValue, setTabValue ] = useState<TabValue>(TabValue.MODELS);

  const handleChangeTabValue = (e: React.SyntheticEvent, tabValue: TabValue) => {
    setTabValue(tabValue);
  }

  return (
    <Tabs 
      value={tabValue} 
      onChange={handleChangeTabValue} 
      textColor={textColor} 
      indicatorColor={indicatorColor}
    >
      {tabList.map((tab: TabI) => (
        <Tab key={tab.value} value={tab.value} label={tab.label} />
      ))}
    </Tabs>
  );
};

export default CommonTabs;