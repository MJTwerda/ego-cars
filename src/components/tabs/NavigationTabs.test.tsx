import { render, screen, fireEvent, act, waitFor, RenderResult } from "@testing-library/react";
import NavigationTabs from "./NavigationTabs";
import { TabColors, NavigationTabI } from "./tabs.interfaces";
import { usePathname, useRouter } from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn(),
}));

const mockPush = jest.fn();
(useRouter as jest.Mock).mockReturnValue({ push: mockPush });

const tabList: NavigationTabI[] = [
  { path: "/", label: "Model" },
  { path: "model-sheet", label: "Model Sheet" },
];

describe("NavigationTabs", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should be render correctly with custom colors", async() => {
    (usePathname as jest.Mock).mockReturnValue("/");
    let component: RenderResult;

    await act(() => {
      component = render(
        <NavigationTabs
          tabList={tabList}
          textColor={TabColors.SECONDARY}
          indicatorColor={TabColors.SECONDARY}
        />
      );
    });
    
    await waitFor(() => {
      expect(component).toMatchSnapshot();
    });
  });

  it("should render tabs correctly", async() => {
    (usePathname as jest.Mock).mockReturnValue("/");
    let component: RenderResult;

    await act(() => {
      component = render(<NavigationTabs tabList={tabList} />);
    })
    
    await waitFor(() => {
      expect(component).toMatchSnapshot();
      tabList.forEach((tab) => {
        const modelTab = component.getByTestId(`navigation-tab-${tab.path}`);
        expect(modelTab).toBeInTheDocument();
      });
    })

  });

  it("should highlight the active tab based on pathname", async() => {
    (usePathname as jest.Mock).mockReturnValue("/model-sheet");
    let component: RenderResult;

    await act(() => {
      component = render(<NavigationTabs tabList={tabList} />);
    });

    await waitFor(() => {
      const activeTab = screen.getByRole("tab", { name: "Model Sheet" });
      expect(activeTab).toHaveAttribute("aria-selected", "true");
    });

  });

  it("should navigate when clicking a tab", async() => {
    (usePathname as jest.Mock).mockReturnValue("/");
    let component: RenderResult;

    await act(() => {
      component = render(<NavigationTabs tabList={tabList} />);
    });

    await waitFor(() => {
      const modelSheetTab = screen.getByRole("tab", { name: "Model Sheet" });
      fireEvent.click(modelSheetTab);
  
      expect(mockPush).toHaveBeenCalledWith("/model-sheet");
    });
  });

  it("should set default tab when pathname does not match", async() => {
    (usePathname as jest.Mock).mockReturnValue("/unknown");
    let component: RenderResult;

    await act(() => {
      component = render(<NavigationTabs tabList={tabList} />);
    });

    await waitFor(() => {
      const defaultTab = screen.getByRole("tab", { name: "Model" });
      expect(defaultTab).toBeInTheDocument();
    });

  });
});