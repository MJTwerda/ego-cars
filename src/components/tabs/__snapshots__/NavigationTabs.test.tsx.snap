// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationTabs should be render correctly with custom colors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTabs-root css-jhczcq-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            class="MuiTabs-flexContainer css-1wxkzlj-MuiTabs-flexContainer"
            role="tablist"
          >
            <a
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-16iolqj-MuiButtonBase-root-MuiTab-root"
              data-testid="navigation-tab-/"
              id="navigation-tab-/"
              role="tab"
              tabindex="0"
            >
              Model
            </a>
            <a
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-16iolqj-MuiButtonBase-root-MuiTab-root"
              data-testid="navigation-tab-model-sheet"
              id="navigation-tab-model-sheet"
              role="tab"
              tabindex="-1"
            >
              Model Sheet
            </a>
          </div>
          <span
            class="MuiTabs-indicator css-ehhs8s-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer css-1wxkzlj-MuiTabs-flexContainer"
          role="tablist"
        >
          <a
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary Mui-selected css-16iolqj-MuiButtonBase-root-MuiTab-root"
            data-testid="navigation-tab-/"
            id="navigation-tab-/"
            role="tab"
            tabindex="0"
          >
            Model
          </a>
          <a
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorSecondary css-16iolqj-MuiButtonBase-root-MuiTab-root"
            data-testid="navigation-tab-model-sheet"
            id="navigation-tab-model-sheet"
            role="tab"
            tabindex="-1"
          >
            Model Sheet
          </a>
        </div>
        <span
          class="MuiTabs-indicator css-ehhs8s-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`NavigationTabs should render tabs correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTabs-root css-jhczcq-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            class="MuiTabs-flexContainer css-1wxkzlj-MuiTabs-flexContainer"
            role="tablist"
          >
            <a
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1usuzwp-MuiButtonBase-root-MuiTab-root"
              data-testid="navigation-tab-/"
              id="navigation-tab-/"
              role="tab"
              tabindex="0"
            >
              Model
            </a>
            <a
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
              data-testid="navigation-tab-model-sheet"
              id="navigation-tab-model-sheet"
              role="tab"
              tabindex="-1"
            >
              Model Sheet
            </a>
          </div>
          <span
            class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTabs-root css-jhczcq-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer css-1wxkzlj-MuiTabs-flexContainer"
          role="tablist"
        >
          <a
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            data-testid="navigation-tab-/"
            id="navigation-tab-/"
            role="tab"
            tabindex="0"
          >
            Model
          </a>
          <a
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1usuzwp-MuiButtonBase-root-MuiTab-root"
            data-testid="navigation-tab-model-sheet"
            id="navigation-tab-model-sheet"
            role="tab"
            tabindex="-1"
          >
            Model Sheet
          </a>
        </div>
        <span
          class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
